<?xml version="1.0"?>
<launch>

  <param name="use_sim_time" value="true"/>
  <arg name="rviz" default="true" />

  <!--<group if="$(arg rviz)">-->
    <!--<include file="$(find rviz_textured_quads)/launch/map_visualization.xml"/>-->
    <!--<node name="rviz" pkg="rviz" type="rviz" args="-d $(find rviz_textured_quads)/demo.rviz" respawn="true" />-->
  <!--</group>-->

<!--   <node name="rqt_topic" pkg="rqt_topic" type="rqt_topic"/> -->
<!--   <node name="rviz" pkg="rviz" type="rviz" args="-d $(find potential_field_navigation)/config/config.rviz" output="screen" respawn="true"/> -->

  
  <!--Image Potential-->
  <node name="charge_image_selecter_node" pkg="potential_field_navigation" type="image_selecter_node" output="screen" cwd="ROS_HOME">
    <param name="image_publisher_topic" value="/image/charge"/>
    <param name="load_as_blue_channel" value="1"/>
    <param name="convert_to_red_channel" value="0"/>
    <param name="charge_as_current" value="0"/>
  </node>
  <node name="charge_image_to_vectorfield_node" pkg="potential_field_navigation" type="image_to_vectorfield_node" output="screen" cwd="ROS_HOME">
    <param name="image_listener_topic" value="/image/charge"/>
    <param name="potentialfield_publisher_topic" value="/potentialfield/image/charge"/>
    <param name="vectorfield_publisher_topic" value="/vectorfield/image/charge"/>
    <param name="heuristic_abs_min" value="0.2"/> <?Undefined An arbitrary value which indecates the start of value decay?>
    <param name="heuristic_factor" value="1.0"/> <?Maximum vector lengh?>
    <param name="heuristic_apply" value="1"/> <?Use heuristic instead of native vectorfield?>
  </node>
  <node name="charge_vectorfield_to_gridmap_node" pkg="potential_field_navigation" type="vectorfield_to_gridmap_node" output="screen" cwd="ROS_HOME">
    <param name="meter_per_pixel" value="0.003"/>
    <param name="frame_id" value="world"/>
    <param name="vectorfield_listener_topic" value="/vectorfield/image/charge"/>
    <param name="gridmap_publisher_topic" value="/vectorfield/image/gridmap/charge"/>
  </node>

  <!--Image Current-->
  <node name="current_image_selecter_node" pkg="potential_field_navigation" type="image_selecter_node" output="screen" cwd="ROS_HOME">
    <param name="image_publisher_topic" value="/image/current"/>
    <param name="load_as_blue_channel" value="1"/>
    <param name="convert_to_red_channel" value="0"/>
    <param name="charge_as_current" value="1"/>
  </node>
  <node name="current_image_to_vectorfield_node" pkg="potential_field_navigation" type="image_to_vectorfield_node" output="screen" cwd="ROS_HOME">
    <param name="image_listener_topic" value="/image/current"/>
    <param name="potentialfield_publisher_topic" value="/potentialfield/image/current"/>
    <param name="vectorfield_publisher_topic" value="/vectorfield/image/current"/>
    <param name="heuristic_abs_min" value="0.2"/> <?Undefined An arbitrary value which indecates the start of value decay?>
    <param name="heuristic_factor" value="1.0"/> <?Maximum vector lengh?>
    <param name="heuristic_apply" value="1"/> <?Use heuristic instead of native vectorfield?>
  </node>
  <node name="current_vectorfield_to_gridmap_node" pkg="potential_field_navigation" type="vectorfield_to_gridmap_node" output="screen" cwd="ROS_HOME">
    <param name="meter_per_pixel" value="0.003"/>
    <param name="frame_id" value="world"/>
    <param name="vectorfield_listener_topic" value="/vectorfield/image/current"/>
    <param name="gridmap_publisher_topic" value="/vectorfield/image/gridmap/current"/>
  </node>

  <!--Fusion-->
  <node name="vectorfield_merger_node" pkg="potential_field_navigation" type="vectorfield_merger_node" output="screen" cwd="ROS_HOME">
    <param name="field1_listener_topic" value="/vectorfield/image/charge"/>
    <param name="field2_listener_topic" value="/vectorfield/image/current"/>
    <param name="vectorfield_publisher_topic" value="/vectorfield/fused"/>
    <param name="normalize" value="1"/>
  </node>
  <node name="fusion_vectorfield_to_gridmap_node" pkg="potential_field_navigation" type="vectorfield_to_gridmap_node" output="screen" cwd="ROS_HOME">
    <param name="meter_per_pixel" value="0.003"/>
    <param name="frame_id" value="world"/>
    <param name="vectorfield_listener_topic" value="/vectorfield/fused"/>
    <param name="gridmap_publisher_topic" value="/vectorfield/fused/gridmap"/>
  </node>

  <!--Steering-->
  <node name="amiro1_vectorfield_to_kinematic_node" pkg="potential_field_navigation" type="vectorfield_to_kinematic_node" output="screen" cwd="ROS_HOME">
    <param name="meter_per_pixel" value="0.003"/>
    <param name="frame_id" value="world"/>
    <param name="vectorfield_listener_topic" value="/vectorfield/fused"/>
    <param name="amiro_odom_listener_topic" value="/amiro1/odom"/>
    <param name="velocityScale_meterPerSecond" value="0.3"/>
    <param name="angularScale_radPerSecond" value="0.1"/>
  </node>
</launch>
